### PR TITLE
Observe region data from all Geo-IP sources

### LINKED ISSUE #492: BUG: Province name is ignored while doing IP-base geolocation
Please see attached screenshot of wttr and weather.com forecast.
There is no sign or rain here and none is likely at this time of year!

screenshot_20200628_131559
screenshot_20200628_131619

### PR DESCRIPTION
**Pull request by gregdan3**
These changes no longer invalidate the cache, but it caches differently from how it used to. Please observe that.

The cache used to be stored as one of the three following:
city;country
countrycode;country;region;city

The former was custom, retrieved from ipinfo, and the latter was stored directly from ip2location. The new format stores data from all three sources, and matches the existing City, Country formatting but with Region added between. 

With some other new changes, latitude and longitude are included in the cache as well. The final format is the following:

COUNTRY_CODE;COUNTRY;REGION;CITY;LATITUDE;LONGITUDE

These changes are as yet untested, which is why I've marked this as a draft. However, I am reasonably confident in these changes thus far; individual functions in location.py have been tested (those I can get keys for). I have yet to read further than wttr.py to find places where the user's location/requested location matter, so there may be other breaking changes I've yet to identify.

This pull request closes #492

### CODE REVIEW

File: lib/location.py [hunk 1]
```diff
@@ -90,106 +90,113 @@ def geolocator(location):
...
-        location = open(cached, 'r').read().split(';')
-        if len(location) > 3:
-            return location[3], location[1]
-        elif len(location) > 1:
-            return location[0], location[1]
-        else:
-            return location[0], None
+    if os.path.exists(cachefile):
+        try:
+            city, region, country = open(cachefile, 'r').read().split(';')
```
Comment (chubin) [inline lib/location.py:115]: Here I would use this code instead:

```
        if location and ';' in location:
            _, country, region, city = location.split(';')
```

File: lib/location.py [hunk 2]
```diff
@@ -90,106 +90,113 @@ def geolocator(location):
...
+        https://stackoverflow.com/questions/466521/how-many-files-can-i-put-in-a-directory
+    """
+    cachefile = os.path.join(IP2LCACHE, ip_addr)
     if not os.path.exists(IP2LCACHE):
         os.makedirs(IP2LCACHE)
-    with open(cached, 'w') as file:
-        file.write(location[0] + ';' + location[1])
+
+    with open(cachefile, 'w') as file:
+        file.write(location[0] + ';' + location[1] + ';' + location[2])
```
Comment (chubin) [inline lib/location.py:102]: I would save the whole line

File: lib/location.py [hunk 3]
```diff
@@ -90,106 +90,113 @@ def geolocator(location):
...
-    return location
+    if not IPINFO_TOKEN:
+        return None, None, None
+    try:
+        r = requests.get(
+            'https://ipinfo.io/%s/json?token=%s'
+            % (ip_addr, IPINFO_TOKEN))
+        r.raise_for_status()
+        r_json = r.json()
+        city, region, country = r_json["city"], r_json["region"], r_json["country"]
```
Comment (chubin) [inline lib/location.py:152]: I would add an empty field (or the real country code if ipinfo.io delivers one) as the 0 field,
for the compatibility reasons

**Review by chubin — COMMENTED** [review]
See my suggestions in #492

File: lib/location.py (hunk 1 above)
Comment (gregdan3) [reply lib/location.py]: I've made this change, now including latlong. See L119 in new file

File: lib/location.py (hunk 2 above)
Comment (gregdan3) [reply lib/location.py]: We cannot save the whole line, because this version is now a compromise between caching only ip2location format and caching all formats. Instead, the iplocating functions now return a tuple of all relevant data, and this function adapts that tuple into the line you see. It's now somewhat longer to include latlong, though

File: lib/location.py (hunk 3 above)
Comment (gregdan3) [reply lib/location.py]: All iplocating functions now return the country code. `ipinfo` in particular only returned 2 character country codes, so I had to add `pycountry` to correct for that, getting the real country name.

File: lib/location.py [hunk 4]
```diff
@@ -125,54 +130,54 @@ def ip2location(ip_addr):
     # if IP2LOCATION_KEY is not set, do not query,
     # because the query wont be processed anyway
     if not IP2LOCATION_KEY:
-        return None, None, None
+        return None
     try:
         r = requests.get(
-            'http://api.ip2location.com/?ip=%s&key=%s&package=WS3'
+            'http://api.ip2location.com/?ip=%s&key=%s&package=WS5'
```
Comment (chubin) [inline lib/location.py:139]: WS5 is 50% more expensive, and is useless for us, because we already have location -> lat,long conversion implementation

Comment (gregdan3) [reply lib/location.py:139]: This means that ip2location will not be able to cache the lat,long at this step, and we'd have to cache differently. How else do you suggest we get that information?

Comment (chubin) [reply lib/location.py:139]: ip2location makes only one step in the conversion pipeline:
it converts IP-address into "location" (a tuple city,province,country),
which is then converted into a (lat,long) pair with `geolocator` (external service).



If we would process only ip addresses, we could stay with this approach,
but we need to convert location queries too (wttr.in/London),
so we need the `geolocator` step anyway

**Comment by chubin** [comment]
@gregdan3 Gregory, I've made a minor code cleanup and extended the doc string of the module.
Don't forget do update your code

I believe, we should settle the cache format prior to further changes, and specify it in the module docstring (I've already added some specification as the first iteration; feel free to suggest any changes if you think the format should be changed).

File: lib/location.py (hunk 4 above)
Comment (gregdan3) [reply lib/location.py:156]: In that case, we should probably remove `lat` and `long` capture from the other two services, and let the pair be resolved later by `geolocator`. Do you think `geolocator` should have its own cache?

**Comment by gregdan3** [comment]
@chubin I've made the necessary changes. The ipcache will now write latlong if they are provided; the two services that already provide latlong still do so, and ip2location has been reduced back to WS3 so it does not.

I think the cache format is fine as is, but I also believe it's worth trying to reduce the number of services you depend on in a single API call. Since many of the ip location services offer latitude and longitude, it should be reasonable to use that info when available and skip the extra call to `geolocator`. That refactor should come later, however.

**Comment by gregdan3** [comment]
Any updates on this one?

**Comment by chubin** [comment]
@gregdan3 Gregory thank you for your pull-request. I saw it, but I still had no time to review it properly. I hope I will to it in the next time (next days).

I fully agree with you regarding geolocator. It is absolutely needed to be integrated to this module.
Also, there are several other features that must be added to this service because of v3.wttr.in (region-handling; in short: wttr.in should handle regions as it handles single locations now, and it is hard to do in a clean way, until the current location.py is properly refactored).

To summarize: I will review and merge your PR in the next days, and then we will continue with the next steps.

**Comment by chubin** [comment]
I've merged and deployed it. Please test.

Maybe it will be a little bit unstable at the beginning, because new naming scheme will cause a lot of cache misses; to mitigate this effect, I activated the "region" feature only for the United States for the moment, because it is the country where it is especially important to have it. In other countries (maybe except India and some other big countries) it is not so crucial, and even more: in some non-English speaking countries, especially if the alphabet is not Latin based, triple names (location-region-country) resolution seems to work worse than pair-based names (location-country).

Thank you for the very cool feature, Gregory, and Happy New Year!

**Comment by gregdan3** [comment]
@chubin The feature appears to work. I am able to request data for a specific IP and get accurate weather, and then request the same location explicitly and still get the same result.

Glad I could help!

**Comment by chubin** [comment]
Now the next step is to refactor `lib/location.py`, because it grew pretty big.
After that I would like to add some features needed for the v3 view (shapes support etc.) 
and integrate the external service responsible for NAME => GPS conversion
